### PR TITLE
[Breaking change] Remove all use of the language service. Use only the Program.

### LINKED ISSUE #1969: sourceFile duplication for TypedRule
### Bug Report

- __TSLint version__: `master`
- __TypeScript version__: `2.1.4` (the one in `package.json`)
- __Running TSLint via__: `./bin/tslint --test test/rules/my-rule`

#### TypeScript code ~~being linted~~

```ts
export class Rule extends Lint.Rules.TypedRule {
    /* tslint:disable:object-literal-sort-keys */
    public static metadata: Lint.IRuleMetadata = {
        ruleName: "my-rule",
        description: "",
        optionsDescription: "Not configurable.",
        options: null,
        optionExamples: ["true"],
        type: "functionality",
        typescriptOnly: false,
    };
    /* tslint:enable:object-literal-sort-keys */

    public applyWithProgram(sourceFile: ts.SourceFile, langSvc: ts.LanguageService): Lint.RuleFailure[] {
        console.log(sourceFile === langSvc.getProgram().getSourceFile(sourceFile.fileName));
        return [];
    }
}
```

#### Actual behavior

Prints `false`.

#### Expected behavior

Prints `true`. The source file used for linting should be the same as the one used for the program.
I want to write a rule that uses the type checker to get information about nodes as I recurse. I can't do that if it uses a different AST.

### LINKED ISSUE #2181: Exception thrown when `no-use-before-declare` is enabled.
### Bug Report

- __TSLint version__: `4.2.2`
- __TypeScript version__: `2.2.0 RC`
- __Running TSLint via__: cli and npm ( `tslint --project tsconfig.json -t stylish --type-check -e **/src/types/api.ts`

#### TypeScript code being linted
I have been unable to narrow it down to a single file, any file in my project I check causes the error to throw.

with `tslint.json` configuration:

```json
{
  "extends": [
    "tslint:recommended",
    "tslint-microsoft-contrib/recommended_ruleset"
  ],

  "rulesDirectory": "node_modules/tslint-microsoft-contrib",

  "rules": {
    "quotemark": [true, "single", "jsx-double"],
    "max-classes-per-file": false,
    "interface-name": false,
    "missing-jsdoc": false,
    "no-unused-variable": false,
    "no-default-export": false,
    "no-relative-imports": false,
    "no-empty-interfaces": false,
    "import-name": false,
    "export-name": false,
    "react-tsx-curly-spacing": [ true, "never" ],
    "array-type": [ true, "generic" ],
    "prefer-array-literal": [ true, { "allow-type-parameters": true } ],
    "prefer-for-of": false,
    "no-increment-decrement": false,
    "no-backbone-get-set-outside-model": false,
    "prefer-type-cast": false,
    "no-angle-bracket-type-assertion": true,
    // This should be enabled for beta and release builds
    "no-suspicious-comment": false,
  },
  "jsRules": {
    "quotemark": [true, "single", "jsx-double"],
    "max-classes-per-file": false,
    "no-unused-variable": false
  }
}
```

#### Actual behavior

Exception is thrown:
```
$ ./node_modules/.bin/tslint --project tsconfig.json -t stylish -c tslint.json --type-check src/index.tsx
no-unused-variable is deprecated. Use the tsc compiler options --noUnusedParameters and --noUnusedLocals instead.
C:\Alcidion\Miya\Web\MiyaED\node_modules\tslint\lib\runner.js:117
            throw error;
            ^

Error: Debug Failure. False expression:
    at Object.assert (C:\Alcidion\Miya\Web\MiyaED\node_modules\typescript\lib\typescript.js:3238:23)
    at getDefaultImportName (C:\Alcidion\Miya\Web\MiyaED\node_modules\typescript\lib\typescript.js:70705:26)
    at Object.getReferencedSymbolsForNode (C:\Alcidion\Miya\Web\MiyaED\node_modules\typescript\lib\typescript.js:70634:51)
    at getSemanticDocumentHighlights (C:\Alcidion\Miya\Web\MiyaED\node_modules\typescript\lib\typescript.js:69952:58)
    at Object.getDocumentHighlights (C:\Alcidion\Miya\Web\MiyaED\node_modules\typescript\lib\typescript.js:69939:29)
    at Object.getDocumentHighlights (C:\Alcidion\Miya\Web\MiyaED\node_modules\typescript\lib\typescript.js:80730:42)
    at NoUseBeforeDeclareWalker.validateUsageForVariable (C:\Alcidion\Miya\Web\MiyaED\node_modules\tslint\lib\rules\noUseBeforeDeclareRule.js:113:47)
    at NoUseBeforeDeclareWalker.visitImportDeclaration (C:\Alcidion\Miya\Web\MiyaED\node_modules\tslint\lib\rules\noUseBeforeDeclareRule.js:77:18)
    at NoUseBeforeDeclareWalker.SyntaxWalker.visitNode (C:\Alcidion\Miya\Web\MiyaED\node_modules\tslint\lib\language\walker\syntaxWalker.js:382:22)
    at NoUseBeforeDeclareWalker.ScopeAwareRuleWalker.visitNode (C:\Alcidion\Miya\Web\MiyaED\node_modules\tslint\lib\language\walker\scopeAwareRuleWalker.js:58:36)
```

#### Expected behavior

Not to throw an error.

### LINKED ISSUE #2210: Wrong no-use-before-declare error
### Bug Report

- __TSLint version__: `"~4.4.2"`
- __TypeScript version__: `2.1.5`
- __Running TSLint via__: Visual Studio Code

#### TypeScript code being linted

```ts
class Greeter {
    greet(options: { str: string }) {
        var {str} = options;
        return str;
    }
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "ban": [true,
        ["_", "extend"],
        ["_", "isNull"],
        ["_", "isDefined"]
    ],
    "class-name": true,
    "curly": true,
    "eofline": false,
    "forin": false,
    "indent": [true, 4],
    "interface-name": true,
    "jsdoc-format": true,
    "label-position": true,
    "max-line-length": [true, 250],
    "no-arg": true,
    "no-bitwise": true,
    "no-console": [true,
        "debug",
        "info",
        "time",
        "timeEnd",
        "trace"
    ],
    "no-construct": true,
    "no-debugger": true,
    "no-duplicate-variable": true,
    "no-empty": false,
    "no-eval": true,
    "no-string-literal": false,
    "trailing-comma": true,
    "no-trailing-whitespace": true,
    "no-unused-expression": true,
    "no-unused-variable": true,
    "no-use-before-declare": true,
    "one-line": [true,
        "check-open-brace",
        "check-catch",
        "check-whitespace"
    ],
    "quotemark": [true, "double"],
    "radix": true,
    "semicolon": true,
    "triple-equals": [true, "allow-null-check"],
    "typedef": [true,
        "indexSignature",
        "propertySignature"
    ],
    "typedef-whitespace": [true,
        ["callSignature", "noSpace"],
        ["catchClause", "noSpace"],
        ["indexSignature", "space"]
    ],
    "variable-name": false,
    "whitespace": [false,
        "check-branch",
        "check-decl",
        "check-operator",
        "check-separator",
        "check-type"
    ]
  }
}```

#### Actual behavior
message: 'variable 'str' used before declaration (no-use-before-declare)'at: '2,22'source: 'tslint'

#### Expected behavior

no error

### PR DESCRIPTION
**Pull request by andy-ms**
#### PR checklist

- [ ] Addresses an existing issue: #1969
  - Should fix #2181 (that error was in document highlights, which we no longer call)
  - Fixes #2210
- [X] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### What changes did you make?

Tslint now only provides the `Program`, not the `LanguageService`. `LanguageService` is for editors; tslint should use the `Program` and `TypeChecker` API.
This would be break any custom typed rules, so should be in a major version change. So long as the custom typed rules do not use the language service, conversion is simple. See [await-promise](https://github.com/palantir/tslint/pull/2235/files#diff-f598bb4bf89a3fbf85ddc367281087f2) for an example.

Only `no-use-before-declare` and `no-unused-variable` used the language service.
Rewrote `no-use-before-declare`. It now gets the symbol at each use and checks that it comes after the declaration.
I think `no-unused-variable` was going to be deprecated anyway, so removed it. It would be possible to rewrite that too if you want to keep it.

As a side-effect, fixes #1969 since we don't have `wrapProgram`, which creates a new language service, which then has its own program. This probably helps performance too, but I haven't tested.

Also changed tslint's own linting to use the program. Using typed rules in this repo could be done in another PR.

Will rebase after #2234 is in.

### CODE REVIEW

File: src/language/rule/typedRule.ts [hunk 1]
```diff
@@ -31,5 +31,5 @@ export abstract class TypedRule extends AbstractRule {
         throw new Error(`The '${this.ruleName}' rule requires type checking`);
     }
 
-    public abstract applyWithProgram(sourceFile: ts.SourceFile, languageService: ts.LanguageService): RuleFailure[];
+    public abstract applyWithProgram(sourceFile: ts.SourceFile, languageService: ts.Program): RuleFailure[];
```
Comment (ajafff) [inline src/language/rule/typedRule.ts]: Also rename the parameter

**Comment by nchen63** [comment]
actually, #1481 is no longer being deprecated due to community response

**Comment by nchen63** [comment]
The parameter originally used `program`, and was changed to use language service in #1608 

Need to verify that the original issue is doesn't break

**Comment by andy-ms** [comment]
Rewrote `no-unused-variable` to not need the language service. Done by creating a duplicate program with the unused-checking compiler options set. Added workaround for Microsoft/TypeScript#9944 that doesn't require `checkEdit`.
This rule would presumably be unnecessary if we fixed that bug and implemented Microsoft/TypeScript#11051.

**Comment by andy-ms** [comment]
Note that this makes `no-unused-variable` and `no-use-before-declare` typed rules, while before they worked without a type checker.

**Comment by nchen63** [comment]
That might be a problem for people who use plug-ins like the VSCode-TSLint since typed rules are not supported. People seem to use this rule to see the warnings instead of using the compiler flag so it doesn't block compilation; see https://github.com/palantir/tslint/issues/1481#issuecomment-262498483.

**Comment by ajafff** [comment]
I'm already working on a solution to use both rules without type checking.
`no-use-before-declare` is no problem.
`no-unused-variable` would only find unused private class members with type checking enabled. The problem with implicit required imports may also still be there - haven't looked at the code yet - so we would exclude them from the check without type checking. This way we would at least get most of the failures and find the rest with type checking.

**Comment by andy-ms** [comment]
It may be possible to create a program consisting of a single file and run the rest as usual.

**Pull request merged**